src/wasm-exts/customModuleDecode.ml

```diff
@@ -684,6 +684,41 @@ let custom_section (name_pred : int list -> bool) (f : int -> stream -> 'a) (def
   | _ -> default
 
 
+let icp_name suffix =
+  let public_name = Utf8.decode ("icp:public " ^ suffix) in
+  let private_name = Utf8.decode ("icp:private " ^ suffix) in
+  fun name ->
+    if public_name = name then
+      Some true
+    else if private_name = name then
+      Some false
+    else None
+
+let icp_custom_section n (f : int -> stream -> 'a) (default : (bool * 'a) option) (s : stream) =
+  let p = pos s in
+  match id s with
+  | Some `CustomSection ->
+    ignore (u8 s);
+    let sec_size = len32 s in
+    let sec_start = pos s in
+    let sec_end = sec_start + sec_size in
+    let name = name s in
+    let opt = icp_name n name in
+    begin
+    match opt with
+    | Some b ->
+      (* this is the right custom section *)
+      let x = f sec_end s in
+      require (pos s = sec_end) s sec_start "custom section size mismatch";
+      Some (b, x)
+    | None -> begin (* wrong custom section, rewind *)
+      s.pos := p;
```

Comment:
Hm, this rewind is a bit ugly (you could start a new stream, skipping over sections isn't that costly). Could you at least abstract that into a function at the top of the module?

Reply:
Done; and added a todo to redo the work for all custom sections. I was merely adapting the custom_section code already present in this file, but I think we should probably just read all custom sections in the manner you suggest, rather than being clever about it (as Joachim was trying to be for the dylink and names sections).